Add Model Factory and Inference Factory

Corsair-cxs opened the pull request:
Add model factory and inference factory, object detection TEST passed.

LewisLiuPub commented:
wonderful design!

Which version of OpenVINO Toolkit did you test based on?
What kinds of tests have you done? 
How many models have you tested?

LewisLiuPub commented:
@Corsair-cxs 
and,

- looks like inference_factory and model_factory can share the same base class, by adding another template type.
- is it possible to add a factory for output classes?

Corsair-cxs commented:
Yes, that's exactly what I'm trying to do.

Corsair-cxs commented:
It's based on ROS Noetic, master branch.

LewisLiuPub commented:
@Corsair-cxs do you have a clear plan to make an update in 1 or 2 weeks? 
If no, I would like to merge this PR, then trigger efforts to upgrade Openvino to 2022.1.
After my upgrading to OV2022.1, you may continue your updating work.

Corsair-cxs commented:
> @Corsair-cxs do you have a clear plan to make an update in 1 or 2 weeks? If no, I would like to merge this PR, then trigger efforts to upgrade Openvino to 2022.1. After my upgrading to OV2022.1, you may continue your updating work.

I have modified the code. 
While I'm trying to add Output Factory, the terminal did not show the image of outputs. So I've not push the latest code for output factory. Maybe we need some time to debug it and find the result.

Corsair-cxs commented:
> > @Corsair-cxs do you have a clear plan to make an update in 1 or 2 weeks? If no, I would like to merge this PR, then trigger efforts to upgrade Openvino to 2022.1. After my upgrading to OV2022.1, you may continue your updating work.
> 
> I have modified the code. While I'm trying to add Output Factory, the terminal did not show the image of outputs. So I've not push the latest code for output factory. Maybe we need some time to debug it and find the result.

The related modify is <https://github.com/Corsair-cxs/ros_openvino_toolkit/commit/cd850c6d1ab6b2cb905474c9461746a8d0370637>

Corsair-cxs commented:
> > @Corsair-cxs do you have a clear plan to make an update in 1 or 2 weeks? If no, I would like to merge this PR, then trigger efforts to upgrade Openvino to 2022.1. After my upgrading to OV2022.1, you may continue your updating work.
> 
> I have modified the code. While I'm trying to add Output Factory, the terminal did not show the image of outputs. So I've not push the latest code for output factory. Maybe we need some time to debug it and find the result.

Hi, Mr.Liu.  I've fixed the bug. 

As your [suggestion](https://github.com/intel/ros_openvino_toolkit/pull/194#issuecomment-1281959639), all of the pipeline modules can be registerd by `std::string`.

LewisLiuPub commented:
Thanks a ton, @Corsair-cxs.
I will review and merge it _after_ another branch's development (upgrading Openvino to API2.0 for Melodic and Noetic).
Will ping you for any comment.
